hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/BlobRenameHandler.java

```diff
@@ -0,0 +1,452 @@
+         * are renamed. When the HBase master renames a log file folder, the lease
+         * locks out other writers.  This prevents a region server that the master
+         * thinks is dead, but is still alive, from committing additional updates.
+         * This is different than when HBase runs on HDFS, where the region server
+         * recovers the lease on a log file, to gain exclusive access to it, before
+         * it splits it.
+         */
+        srcAbfsLease = takeLease(src, srcEtag);
+        srcLeaseId = srcAbfsLease.getLeaseID();
+        if (!isAtomicRenameRecovery && pathInformation.getIsDirectory()) {
+          /*
+           * if it is not a resume of a previous failed atomic rename operation,
+           * perform the pre-rename operation.
+           */
+          renameAtomicity = getRenameAtomicity(pathInformation);
+          renameAtomicity.preRename();
+        }
+      }
+      if (pathInformation.getIsDirectory()) {
+        result = listRecursiveAndTakeAction() && finalSrcRename();
+      } else {
+        result = renameInternal(src, dst);
+      }
+      if (result && renameAtomicity != null) {
+        renameAtomicity.postRename();
+      }
+      return new AbfsClientRenameResult(null, result, false);
+    } else {
+      return new AbfsClientRenameResult(null, false, false);
+    }
+  }
+
+  private boolean finalSrcRename() throws IOException {
+    tracingContext.setOperatedBlobCount(operatedBlobCount.get() + 1);
+    try {
+      return renameInternal(src, dst);
+    } finally {
+      tracingContext.setOperatedBlobCount(null);
+    }
+  }
+
+  @VisibleForTesting
+  public RenameAtomicity getRenameAtomicity(final PathInformation pathInformation)
+      throws IOException {
+    return new RenameAtomicity(src,
+        dst,
+        new Path(src.getParent(), src.getName() + RenameAtomicity.SUFFIX),
+        tracingContext,
+        pathInformation.getETag(),
+        abfsClient);
+  }
+
+  private AbfsLease takeLease(final Path path, final String eTag)
+      throws AzureBlobFileSystemException {
+    AbfsLease lease = new AbfsLease(abfsClient, path.toUri().getPath(),
+        abfsClient.getAbfsConfiguration()
+            .getAtomicRenameLeaseRefreshDuration(),
+        eTag,
+        tracingContext);
+    leases.add(lease);
+    return lease;
+  }
+
+  private boolean containsColon(Path p) {
+    return p.toUri().getPath().contains(":");
+  }
+
+  /**
+   * Since, server doesn't have a rename API and would not be able to check HDFS
+   * contracts, client would have to ensure that no HDFS contract is violated.
+   *
+   * @param src source path
+   * @param dst destination path
+   * @param pathInformation object in which path information of the source path would be stored
+   *
+   * @return true if the pre-checks pass
+   * @throws AzureBlobFileSystemException if server call fails or given paths are invalid.
+   */
+  private boolean preCheck(final Path src, final Path dst,
```

Comment:
Method is too big, can be broken down into smaller chunks like :-  validateDestinationPath(dst);
    validateDestinationIsNotSubDir(src, dst);
    validateSourcePath(src, pathInformation);
    validateEtag(srcEtag, pathInformation.getETag());
    validateDestinationNameForDirectory(src, dst, pathInformation);
    validateDestinationParentExists(dst, pathInformation);

Reply:
taken.